### PR TITLE
Control, Shift and Arrow Key functionality for Chromebook

### LINKED ISSUE #15758: Physical keyboard: text selection does not work with ctrl + shift modifiers
Modifiers combined with arrow keys should influence text selection.

### PR DESCRIPTION
**Pull request by jslavitz**
fixes #15758

Added control, shift and arrow key text manipulation to keyboards on the chromebook and bluetooth keyboards connected to android devices. The up and down arrows now function correctly in respect to a text field moving the cursor up and down as expected.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -198,6 +199,120 @@ class RenderEditable extends RenderBox {
 
   Rect _lastCaretRect;
 
+  bool _shiftIsDown = false;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: Won't these have the same issue that shift has?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -815,4 +936,4 @@ class RenderEditable extends RenderBox {
       ),
     ];
   }
-}
+}
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: nit: newline at EOF

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 3]
```diff
@@ -233,7 +239,20 @@ class RawKeyboard {
     _listeners.remove(listener);
   }
 
+  final int _leftShiftCode = 59;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/raw_keyboard.dart]: Same as above, make these const

File: packages/flutter/test/material/text_field_test.dart [hunk 4]
```diff
@@ -1764,6 +1764,177 @@ void main() {
...
+            ),
+          ),
+        ),
+      );
+    }
+
+    testWidgets('Shift test 1', (WidgetTester tester) async{
+
+      await tester.pumpWidget(setupWidget());
+      const String testValue = 'your mom';
```
Comment (jonahwilliams) [inline packages/flutter/test/material/text_field_test.dart]: maybe choose a different string to test

File: packages/flutter/test/material/text_field_test.dart [hunk 5]
```diff
@@ -1764,6 +1764,177 @@ void main() {
...
+    testWidgets('Control Shift test', (WidgetTester tester) async{
+      await tester.pumpWidget(setupWidget());
+      const String testValue = 'your mom';
+      await tester.enterText(find.byType(TextField), testValue);
+
+      await tester.idle();
+      await tester.tap(find.byType(TextField));
+      await tester.pumpAndSettle();
+
+      sendKeyEventWithCode(59);
```
Comment (jonahwilliams) [inline packages/flutter/test/material/text_field_test.dart]: leave comments here describing what they do:

`sendKeyEventWithCode(59); //  NAME key down event

**Review by jonahwilliams — DISMISSED** [review]
Left comments, also the _keyEvent method is hard to follow. Consider ways to reduce the amount of nested ifs and documentation to make it more obvious what decisions you are making

File: packages/flutter/lib/src/rendering/editable.dart [hunk 6]
```diff
@@ -198,6 +199,120 @@ class RenderEditable extends RenderBox {
...
+      if (_downArrowDown || _upArrowDown) {
+        double verticalOffset = _textPainter.preferredLineHeight;
+        if (_upArrowDown)
+          verticalOffset *= -1;
+
+        final Offset caretOffset = _textPainter.getOffsetForCaret(new TextPosition(offset: _extentOffset), _caretPrototype);
+        final Offset caretOffsetTranslated = caretOffset.translate(0.0, verticalOffset);
+        final TextPosition position = _textPainter.getPositionForOffset(caretOffsetTranslated);
+
+        if (position.offset == _extentOffset){
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: How should I do that? I'm pretty sure I don't have more conditionals than I need?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: What issue is that? I think it should be fine. I check for the shift key being pressed with every key press so this variable should always represent the state of the shift key.

File: packages/flutter/test/material/text_field_test.dart (hunk 4 above)
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: hahahaha, fair enough

File: packages/flutter/lib/src/rendering/editable.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/editable.dart]: The example I gave should be equivalent to L280-L287

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/editable.dart]: Oh I see that now, perfect. Do you still need to track all of the keystate here if it is in Keyboard too?

Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: Yah, wait I don't because now I'm updating the value for shift with every key press.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 7]
```diff
@@ -198,6 +199,138 @@ class RenderEditable extends RenderBox {
...
+  static const int _kUpArrowCode = 19;
+  static const int _kDownArrowCode = 20;
+
+  int _extentOffset = -1;
+  int _baseOffset = -1;
+
+  int _previousCursorLocation = -1;
+  bool _resetCursor = false;
+
+  void _keyEvent(RawKeyEvent keyEvent){
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: Thank you for adding the documentation - it is a lot easier to follow what is going on here.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 8]
```diff
@@ -198,6 +199,138 @@ class RenderEditable extends RenderBox {
...
+    if (!(keyEvent.data is RawKeyEventDataAndroid))
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: I _think_ it would be better to check `defaultTargetplatform` is `Platform.android`, but I'm not sure here.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 2 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: I don't think that's a new line? Looks like I just added it and deleted it.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 9]
```diff
@@ -198,6 +199,138 @@ class RenderEditable extends RenderBox {
...
+    final RawKeyEventDataAndroid rawAndroidEvent = keyEvent.data;
+    final int pressedKeyCode = rawAndroidEvent.keyCode;
+
+    if (selection.isCollapsed) {
+      _extentOffset = selection.extentOffset;
+      _baseOffset = selection.baseOffset;
+    }
+
+    // Update current key states
+    final bool pressedDown = keyEvent is RawKeyDownEvent;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: Do we actually need to do anything for keyup? If so, is the logic for keyDown and keyUp disjoint enough that we could separate it into different methods?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 7 above)
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: Ultra nit: name this method `_handleKeyEvent`

File: packages/flutter/lib/src/rendering/editable.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/editable.dart]: I mean - nit, add the newline at EOF :)

**Comment by Hixie** [comment]
Apparently KeyEvent.isShiftPressed() is the way on Android to tell if a shift key is pressed. We should use that rather than globally tracking the up/down state, so that we can handle things like the shift key being released while the app isn't focused.

**Comment by jonahwilliams** [comment]
Thankfully it looks like there is a`isCtrlPressed()` for the control key too. @jslavitz this will take a small change to the method [here](https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/view/FlutterView.java#L195) to add the control and shift down state.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 9 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: I mean, all key up does is change the value of the arrow down booleans. Because the conditional will only execute on key down.

**Comment by jslavitz** [comment]
@Hixie  We need to expose that value in the engine tho correct?

**Comment by jonahwilliams** [comment]
It sounds like the current shift/control key state is also exposed via masks like https://developer.android.com/reference/android/view/KeyEvent#META_SHIFT_MASK, we might already have the information

File: packages/flutter/lib/src/rendering/editable.dart [hunk 10]
```diff
@@ -198,6 +198,142 @@ class RenderEditable extends RenderBox {
...
+  int _baseOffset = -1;
+
+  int _previousCursorLocation = -1;
+  bool _resetCursor = false;
+
+  static const int _kShiftMask = 1;
+  static const int _kControlMask = 1 << 12;
+
+  void _handleKeyEvent(RawKeyEvent keyEvent){
+    if(defaultTargetPlatform != TargetPlatform.android)
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: nit: space after if

Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: though maybe this would be better as a switch, where the iOS branch says that this is handled by the ObjectiveC code and the Fuchsia branch has a TODO?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 11]
```diff
@@ -465,6 +606,7 @@ class RenderEditable extends RenderBox {
   void detach() {
     _offset.removeListener(markNeedsPaint);
     _showCursor.removeListener(markNeedsPaint);
+    RawKeyboard.instance.removeListener(_handleKeyEvent);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: this should presumably only happen if we're focused

Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: attach also needs to be handled

File: packages/flutter/test/material/text_field_test.dart [hunk 12]
```diff
@@ -1764,6 +1764,164 @@ void main() {
...
+
+      sendKeyEventWithCode(19, true, true, false);               // UP_ARROW keydown
+      await tester.pumpAndSettle();
+      sendKeyEventWithCode(19, false, true, false);              // UP_ARROW keyup
+      await tester.pumpAndSettle();
+
+      expect(controller.selection.extentOffset - controller.selection.baseOffset, 5);
+    });
+  });
+
```
Comment (Hixie) [inline packages/flutter/test/material/text_field_test.dart]: some other tests to have would be:
- if you have two text fields, and both think they are focused, do they both get the events?
- if you have two text fields, and you switch the focus from one to the other and back again, does only the focused one react?
- if you have a text field and you move it from one part of the tree to another using a GlobalKey, does it keep receiving events after?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 11 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: So right now I'm adding and removing based on focus. So this line will only execute if the text field isn't unfocused, but application focus changes. So it's definitely possible this executes even when the listener has already been removed. Which I think is fine?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 13]
```diff
@@ -198,6 +198,142 @@ class RenderEditable extends RenderBox {
 
   Rect _lastCaretRect;
 
+  bool _rightArrowDown = false;
+  bool _leftArrowDown = false;
+  bool _upArrowDown = false;
+  bool _downArrowDown = false;
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: In case a user focuses when a key is already held down. Do I not need to worry about this case?

File: packages/flutter/test/material/text_field_test.dart [hunk 14]
```diff
@@ -1764,6 +1764,164 @@ void main() {
...
+      'keymap': 'android',
+      'keyCode' : code,
+      'hidUsage': 0x04,
+      'codePoint': 0x64,
+      'metaState': metaState,
+    });
+  }
+
+  group('Keyboard Tests', (){
+    List<RawKeyEvent> events;
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: I do? It's used for onKey.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: How does it help with that case? Can you elaborate?
Also that would be a good test to add.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: We shouldn't remove listeners that aren't added.

File: packages/flutter/test/material/text_field_test.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/test/material/text_field_test.dart]: onKey adds to it, but unless I'm missing something, you never read from it.

File: packages/flutter/test/material/text_field_test.dart (hunk 12 above)
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: 1) That should work right? They should both get events, correct?

File: packages/flutter/test/material/text_field_test.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: onkey is required tho... should I have it do something else instead then?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: Ok, yah I think you're right actually. Because we get a function call every frame for the button that's been pressed most recently. And you can only really move the cursor in one direction at once (at least on typical keyboards). So I'll make them locals then...

File: packages/flutter/test/material/text_field_test.dart (hunk 12 above)
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: / how should I focus both fields? I can't use the tester.tap, because that will only focus one of them...

Comment (Hixie) [reply packages/flutter/test/material/text_field_test.dart]: Right, if they both think they're focused they should both get events.
One way to test it would be to have a Column with two MaterialApps inside. Each MaterialApp will starts its own independent focus scope, I believe.

File: packages/flutter/test/material/text_field_test.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/test/material/text_field_test.dart]: If you're not using it at all, then I guess the widget isn't needed either? I'm not sure what you were trying to do here.

[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: But onKey is required in the constructor?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 11 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: So how can I account for the case where the editable is detached, but not unfocused? Or will that always be the case?

File: packages/flutter/test/material/text_field_test.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/test/material/text_field_test.dart]: Ok let me back up a bit. Why do you need the constructor?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: If it's detached, it can't be focused, because it's not in the tree (and thus not on the screen).

File: packages/flutter/test/material/text_field_test.dart (hunk 12 above)
Comment (jslavitz) [reply packages/flutter/test/material/text_field_test.dart]: Also for 2), what are we trying to test here? What do you mean react? Obviously typing works correctly in this situation.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: "detach" basically is the equivalent of "dispose", except that it's possible for it to be "attach"ed again afterwards. RenderObjects start in the "detached" state.

File: packages/flutter/test/material/text_field_test.dart (hunk 12 above)
Comment (Hixie) [reply packages/flutter/test/material/text_field_test.dart]: The purpose of tests is to make sure we don't accidentally break things later, not to verify that it works today. I can easily imagine scenarios where we accidentally break things so that switching focus no longer correctly results in the right text field detecting the keystrokes.

Comment (jonahwilliams) [reply packages/flutter/test/material/text_field_test.dart]: @Hixie Jeremy and I took a look at this today and we've come to the conclusion that there is no way to set this up in our current test environment. The test binding has a single slot for focus which it maintains, so building two material apps causes the second to entirely steal focus for the first.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 15]
```diff
@@ -301,10 +447,16 @@ class RenderEditable extends RenderBox {
   bool get hasFocus => _hasFocus;
   bool _hasFocus;
   set hasFocus(bool value) {
+//    debugPrint(value.toString() + " and " + this.hashCode.toString());
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: Remove this line

File: packages/flutter/lib/src/rendering/editable.dart [hunk 16]
```diff
@@ -198,6 +198,152 @@ class RenderEditable extends RenderBox {
...
+    final bool leftArrow = pressedKeyCode == _kLeftArrowCode;
+    final bool upArrow = pressedKeyCode == _kUpArrowCode;
+    final bool downArrow = pressedKeyCode == _kDownArrowCode;
+    final bool arrow = leftArrow || rightArrow || upArrow || downArrow;
+
+    // We will only move select or more the caret if an arrow is pressed
+    if (arrow) {
+      int newOffset = _extentOffset;
+
+      newOffset = _handleControl(rightArrow, leftArrow, ctrl, newOffset);
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: Instead of checking for the condition in each handleMethod, you should do it up front and only pass the needed values

File: packages/flutter/lib/src/rendering/editable.dart [hunk 17]
```diff
@@ -198,6 +198,152 @@ class RenderEditable extends RenderBox {
...
+      newOffset = _handleHorizontalArrows(rightArrow, leftArrow, shift, newOffset);
+      newOffset = _handleVerticalArrows(upArrow, downArrow, shift, newOffset);
+      newOffset = _handleSelection(rightArrow, leftArrow, shift, newOffset);
+
+      _extentOffset = newOffset;
+    }
+  }
+
+  int _handleSelection(bool rightArrow, bool leftArrow, bool shift, int newOffset) {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/editable.dart]: It would be great if you gave each of these a short description of what they might do to text selection

File: packages/flutter/lib/src/rendering/editable.dart [hunk 18]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+
+      _extentOffset = newOffset;
+    }
+  }
+
+  // Handles the final selection of text or removal of the selection and placing
+  // of the caret
+  int _handleShift(bool rightArrow, bool leftArrow, bool shift, int newOffset) {
+    // For some reason, deletion only works if the base offset is less
+    // than the extent offset.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: we should definitely figure out the reason before we land this.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 19]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+    if (shift) {
+      if (_baseOffset < newOffset)
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: nit: please use braces around if blocks if they're more than one line long

File: packages/flutter/lib/src/rendering/editable.dart [hunk 20]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+        onSelectionChanged(new TextSelection(
+          baseOffset: _baseOffset, extentOffset: newOffset),
+          this, SelectionChangedCause.keyboard);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: also generally please use flutter indent style, as in:

```dart
        onSelectionChanged(
          new TextSelection(
￼            baseOffset: _baseOffset,
            extentOffset: newOffset,
          ),
￼          this,
          SelectionChangedCause.keyboard,
        );
```

File: packages/flutter/lib/src/rendering/editable.dart [hunk 21]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+  }
+
+  // Handles the final selection of text or removal of the selection and placing
+  // of the caret
+  int _handleShift(bool rightArrow, bool leftArrow, bool shift, int newOffset) {
+    // For some reason, deletion only works if the base offset is less
+    // than the extent offset.
+    if (shift) {
+      if (_baseOffset < newOffset)
+        onSelectionChanged(new TextSelection(
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: what if onSelectionChanged is null?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 22]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+      newOffset = _handleControl(rightArrow, leftArrow, ctrl, newOffset);
+      newOffset = _handleHorizontalArrows(rightArrow, leftArrow, shift, newOffset);
+      newOffset = _handleVerticalArrows(upArrow, downArrow, shift, newOffset);
+      newOffset = _handleShift(rightArrow, leftArrow, shift, newOffset);
+
+      _extentOffset = newOffset;
+    }
+  }
+
+  // Handles the final selection of text or removal of the selection and placing
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: final?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 23]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+  // of the caret
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: trivial nit: end comments with a period

File: packages/flutter/lib/src/rendering/editable.dart [hunk 24]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+          baseOffset: _baseOffset, extentOffset: newOffset),
+          this, SelectionChangedCause.keyboard);
+      else
+        onSelectionChanged(new TextSelection(
+          baseOffset: newOffset , extentOffset: _baseOffset),
+          this, SelectionChangedCause.keyboard);
+    }
+    // We want to put the cursor at the correct location depending on which
+    // arrow is used while there is a selection.
+    else {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: trivial nit: put the comment inside the else block so that the else block can use the conventional `} else {` style

File: packages/flutter/lib/src/rendering/editable.dart [hunk 25]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+      newOffset = _handleVerticalArrows(upArrow, downArrow, shift, newOffset);
+      newOffset = _handleShift(rightArrow, leftArrow, shift, newOffset);
+
+      _extentOffset = newOffset;
+    }
+  }
+
+  // Handles the final selection of text or removal of the selection and placing
+  // of the caret
+  int _handleShift(bool rightArrow, bool leftArrow, bool shift, int newOffset) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: nit: please put these methods in the order that they are called, it makes reading the code easier

File: packages/flutter/lib/src/rendering/editable.dart [hunk 26]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+    // For some reason, deletion only works if the base offset is less
+    // than the extent offset.
+    if (shift) {
+      if (_baseOffset < newOffset)
+        onSelectionChanged(new TextSelection(
+          baseOffset: _baseOffset, extentOffset: newOffset),
+          this, SelectionChangedCause.keyboard);
+      else
+        onSelectionChanged(new TextSelection(
+          baseOffset: newOffset , extentOffset: _baseOffset),
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: trivial nit: you have an extra space before a comma

File: packages/flutter/lib/src/rendering/editable.dart [hunk 27]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+    // If control is pressed, we will decide which way to look for a word
+    // based on which arrow is pressed.
+    if (ctrl) {
+      if (leftArrow && _extentOffset > 2) {
+        final TextSelection textSelection = _selectWordAtOffset(new TextPosition(offset: _extentOffset - 2));
+        newOffset = textSelection.baseOffset + 1;
+      } else if (rightArrow && _extentOffset < text.text.length - 2) {
+        final TextSelection textSelection = _selectWordAtOffset(new TextPosition(offset: _extentOffset + 1));
+        newOffset = textSelection.extentOffset - 1;
+      }
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: does this work is the cursor is between the first and second characters of a string?

As in, `a|bc`, press ctrl+left, does it end up at the start?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 28]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+        _previousCursorLocation -= 1;
+    }
+    return newOffset;
+  }
+
+  // Handles full word traversal using control.
+  int _handleControl(bool rightArrow, bool leftArrow, bool ctrl, int newOffset) {
+    // If control is pressed, we will decide which way to look for a word
+    // based on which arrow is pressed.
+    if (ctrl) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: since this method only does something if ctrl is true, it might be clearer (and would certainly be more efficient) to only call it when ctrl is true, and not have it deal with ctrl at all.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 29]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+        _resetCursor = false;
+      } else {
+        newOffset = position.offset;
+        _previousCursorLocation = newOffset;
+      }
+    }
+    return newOffset;
+  }
+
+  // Simply handles the horizontal arrows
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: this comment doesn't add anything, you can remove it

File: packages/flutter/lib/src/rendering/editable.dart [hunk 30]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+      newOffset += 1;
+      if (shift)
+        _previousCursorLocation += 1;
+    }
+    if (leftArrow && _extentOffset > 0) {
+      newOffset -= 1;
+      if (shift)
+        _previousCursorLocation -= 1;
+    }
+    return newOffset;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: we probably need to make sure that left arrow means +1 and right arrow means -1 in TextDirection.rtl text fields.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 31]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+
+      final Offset caretOffset = _textPainter.getOffsetForCaret(new TextPosition(offset: _extentOffset), _caretPrototype);
+      final Offset caretOffsetTranslated = caretOffset.translate(0.0, verticalOffset);
+      final TextPosition position = _textPainter.getPositionForOffset(caretOffsetTranslated);
+
+      // To account for the possibility where the user vertically highlights
+      // all the way to the top or bottom of the text, we hold the previous
+      // cursor location. This allows us to restore to this position in the
+      // case that the user wants to unhighlight some text.
+      if (position.offset == _extentOffset){
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: trivial nit: missing space before `{`

File: packages/flutter/lib/src/rendering/editable.dart [hunk 32]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+  static const int _kLeftArrowCode = 21;
+  static const int _kRightArrowCode = 22;
+  static const int _kUpArrowCode = 19;
+  static const int _kDownArrowCode = 20;
+
+  int _extentOffset = -1;
+  int _baseOffset = -1;
+
+  int _previousCursorLocation = -1;
+  bool _resetCursor = false;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: please document what these private variables are

File: packages/flutter/lib/src/rendering/editable.dart [hunk 33]
```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
...
+        this, SelectionChangedCause.keyboard);
+    }
+    return newOffset;
+  }
+
+  // Handles moving the cursor vertically as well as taking care of the
+  // case where the user moves the cursor to the end or beginning of the text
+  // and then back up or down.
+  int _handleVerticalArrows(bool upArrow, bool downArrow, bool shift, int newOffset) {
+    if (downArrow || upArrow) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: as with ctrl, there's probably no point calling this method at all if neither is true

**Comment by Hixie** [comment]
We should probably add a TODO saying that all this should be changed to work in terms of extended grapheme clusters instead of UTF-16 words. The current logic will fall over completely e.g. in the presence of emoji. Unfortunately Dart doesn't expose extended grapheme clusters yet.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 18 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: This was actually documented in the code somewhere in the TextSelection class, that the text selection is only defined if the the base offset is less than the extent offset.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 21 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: Then nothing works?

Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: we definitely shouldn't crash. we can consider making it required that it be non-null, or we can not support selection modification when it's null, etc.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 30 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: Wait, do I need to take into account rtl text fields? I think it would be easiest to just remap left as right and vice versa.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 33 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: I mean, that's true for every button? I feel like it's cleaner to check for the key presses within the function though I can move them outside if you feel that would be better.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 30 above)
Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: Looks like left and right actually give the correct output even with textdirection.rtl enabled.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 27 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: ah, i see. Maybe add a comment in this function (the one for jumping by word) saying that if you'd only move one character anyway, we handle it by ignoring the ctrl?
What happens if you're two words out and press ctrl-left? do we do both the left jump by word and the move left one character? That's... weird.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 21 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: i haven't checked if it's common to not provide it, or what happens today if you don't.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 30 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: make sure you test with actual hebrew text, otherwise the text is still left to right even though the field is right to left.

Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: (maybe also check what android fields do here. maybe left always means "forward" even in RTL?)

Comment (jslavitz) [reply packages/flutter/lib/src/rendering/editable.dart]: That seems to be the case.

**Comment by jslavitz** [comment]
I accidentally started working on a new fix in the same branch so I had to delete some code... I'll definitely do more self reviewing next time! Thanks for your help on all this though.

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/editable.dart [hunk 34]
```diff
@@ -198,6 +198,186 @@ class RenderEditable extends RenderBox {
...
+  static const int _kUpArrowCode = 19;
+  static const int _kDownArrowCode = 20;
+
+  // The extent offset of the current selection
+  int _extentOffset = -1;
+
+  // The base offset of the current selection
+  int _baseOffset = -1;
+
+  // Holds the last location the user selected in the case that he selects all
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: he select -> they select

File: packages/flutter/lib/src/rendering/editable.dart [hunk 35]
```diff
@@ -198,6 +198,186 @@ class RenderEditable extends RenderBox {
...
+    final bool leftArrow = pressedKeyCode == _kLeftArrowCode;
+    final bool upArrow = pressedKeyCode == _kUpArrowCode;
+    final bool downArrow = pressedKeyCode == _kDownArrowCode;
+    final bool arrow = leftArrow || rightArrow || upArrow || downArrow;
+
+    // We will only move select or more the caret if an arrow is pressed
+    if (arrow) {
+      int newOffset = _extentOffset;
+
+      // Because the user can use multiple keys to change how he selects
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: ditto

File: packages/flutter/lib/src/rendering/editable.dart (hunk 30 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: surprising, but makes life easier for us i guess

**Comment by Hixie** [comment]
LGTM modulo minor nits

File: packages/flutter/lib/src/rendering/editable.dart (hunk 33 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart]: It doesn't make a huge difference in this case, but in the general case you want code to only call methods that they want to actually run, because calling a method itself has a non-trivial cost.
